### PR TITLE
Add best practices for using Rx with ViewModels and Testing

### PR DESCRIPTION
**Pull request by cprime**

### CODE REVIEW

File: ios/rxswift_sherpa.md [hunk 1]
```diff
@@ -1,4 +1,122 @@
...
+init(identifier: String) {
+  APIClient.user(withIdentifier: identifier).catchErrorJustReturn(nil).bind(to: user).disposed(by: disposeBag)
+  username = user.asObservable.map({ $0.username })
+}
+```
+
+The majority of a ViewModel's exposed observable sequences will be observed by objects in the UI layer. CocoaTouch requires that UI updates be performed on the main thread. Additionally, since these observations should generally continue until the ViewController leaves the screen, the observed sequences should not complete or error. For these reasons, you should use the correct observable type for the given situation:
+
+* Use a `Driver` when data is set by the ViewModel and is read-only from the View's perspective
+* Use a `Variable` when double-binding is required to allow the observing View to update observable sequence's most recent value based on user input
```
Comment (vinnyt) [inline ios/rxswift_sherpa.md]: double-binding -> bi-directional

File: ios/rxswift_sherpa.md [hunk 2]
```diff
@@ -1,4 +1,122 @@
...
+* Use of `Observable` should be limited to when the initial stream is a immediately available; although, in these cases, a plain swift property may be sufficient
```
Comment (colinbrash) [inline ios/rxswift_sherpa.md]: Should `username` in the example above be a `Driver` then?

Comment (Tom-Le) [inline ios/rxswift_sherpa.md]: Extra "a" between "is a immediately"

File: ios/rxswift_sherpa.md [hunk 3]
```diff
@@ -1,4 +1,122 @@
...
+
+__Open Question: Should we be advocating against using Observable at all?__
```
Comment (Tom-Le) [inline ios/rxswift_sherpa.md]: If we're talking about in the context of a view model exposing observable sequences to a view controller then yeah, I'd advocate against using `Observable`s at all. Not in general tho; for instance in the example above `APIClient.user(_:)` should absolutely return a normal `Observable`, but `username` should be a `Driver`.

File: ios/rxswift_sherpa.md [hunk 4]
```diff
@@ -1,4 +1,122 @@
...
+```
+
+### Disposing
+
+Each ViewController should be given a `var` DisposeBag for use with outlet bindings and other subscriptions. This bag should be replaced within `viewWillDisappear`, or in a dependent method, and balances against the bindings performed in `viewWillAppear`. This allows the ViewController to disable UI updates while it is not on screen.
+
+```
+override func viewWillDisappear(_ animated: Bool) {
+  super.viewWillDisappear(animated)
+  disposeBag = DisposeBag
```
Comment (Tom-Le) [inline ios/rxswift_sherpa.md]: Missing parentheses next to `DisposeBag`

File: ios/rxswift_sherpa.md (hunk 3 above)
Comment (Tom-Le) [reply ios/rxswift_sherpa.md]: It's probably clearer if we just clarify what a `Driver` is (namely, `Observable` + `.observeOn(MainScheduler.instance)` + `.shareReplayLatestWhileConnected()`), then stipulate that only `Driver`s should be bound to UI elements

File: ios/rxswift_sherpa.md [hunk 5]
```diff
@@ -1,4 +1,122 @@
...
+#### Where to perform the bindings?
+
+A ViewModel's observables should be bound to a ViewController's outlets from `viewWillAppear` or in a dependent method. This allows time for the ViewModel to be injected into the ViewController (if necessary) and for the outlets initial value to be set prior to the View being displayed on screen. Additionally, it allows for the disposal and rebinding when the view controller leaves and returns to the screen.
+
+```
+override func viewWillAppear(_ animated: Bool) {
+  super.viewWillAppear(animated)
+  viewModel.title.bind(to: self.rx.title).disposed(by: disposeBag)
+}
+```
```
Comment (colinbrash) [inline ios/rxswift_sherpa.md:71]: I think we need to explicitly show the disposal as well, or we will end up with a lot of people missing that step.

Comment (Tom-Le) [reply ios/rxswift_sherpa.md:71]: The "Disposing" section down below is probably good enough tho IMO

**Review by Tom-Le — APPROVED** [review]

**Pull request merged**